Add design doc

diptadas opened the pull request.

diptadas committed:
Added initial design doc

diptadas committed:
Removed username flag

nklaassen commented on Design.md:
```diff
@@ -0,0 +1,128 @@
+
+## Job handling
+
+- Wrap the linux command in `bash`
```
Don't wrap the command with `bash` by default, please let the client send the exact command, with arguments. If the client wants to run bash, they can do so explicitly

diptadas replied:
Added a separate field for specifying arguments as a list of strings.

nklaassen commented on Design.md:
```diff
@@ -0,0 +1,128 @@
+- `Created` Received a job creation request (short-lived since we immediately start the job instead of queueing)
+- `Running` Job process started
+- `Running_Timeout` Failed to kill process after timeout
+- `Running_Force_Stop` Failed to kill process after force stop requested
```
How will you force stop jobs, and how can this fail?

diptadas replied:
We can use `Cmd.Process.Kill()` to force stop the process. However, there can be errors while terminating the process and it can be still running. Use `Running_Force_Stop` status to indicate this scenario.

nklaassen commented on Design.md:
```diff
@@ -0,0 +1,128 @@
+
+- `Created` Received a job creation request (short-lived since we immediately start the job instead of queueing)
+- `Running` Job process started
+- `Running_Timeout` Failed to kill process after timeout
```
How will you kill the process after timeout, and how can this fail?

diptadas replied:
Removed the timeout as suggested.

nklaassen commented on Design.md:
```diff
@@ -0,0 +1,128 @@
+- `ID` unique identifier of a job
+- `Command` linux command to be executed
+- `Status` current status of the job
+- `Output` output of the job process
```
Can you describe more what exactly `Output` will be, and how you will handle it?

diptadas replied:
A string that represents the combined stderr and stdout.

nklaassen commented on Design.md:
```diff
@@ -0,0 +1,128 @@
+- `Command` linux command to be executed
+- `Status` current status of the job
+- `Output` output of the job process
+- `Error` exit error or, termination error
```
What will `Error` be? A string, integer, something else?

diptadas replied:
A string

nklaassen commented on Design.md:
```diff
@@ -0,0 +1,128 @@
+
+## API
+
+- `POST /create {"command": "pwd", "timeout": "1"}`
+- `GET /status/{id}`
+- `PUT /stop/{id}`
+- `PUT /purge/{id}`
```
Can you show an example response for these?

diptadas replied:
Added sample requests and responses for each API.

Joerger commented on Design.md:
```diff
@@ -0,0 +1,128 @@
+    - CLI should be able to connect to worker service and schedule several jobs.
+    - CLI should be able to query result of the job execution and fetch its output.
+
+## Job object
```
Can you specify the type for each of the Job object's fields?

diptadas replied:
Updated design doc to specify types for each field.

Joerger commented on Design.md:
```diff
@@ -0,0 +1,128 @@
+- `Status` current status of the job
+- `Output` output of the job process
+- `Error` exit error or, termination error
+- `Timeout` timeout in seconds to stop the job
```
No need to implement a timeout mechanism.

diptadas replied:
Removed

Joerger commented on Design.md:
```diff
@@ -0,0 +1,128 @@
+    - Check the current status of the job
+    - Stop the job process if it is running
+    - Send confirmation on success or, error on termination failure
+- `PurgeJob(jobID)`
```
This isn't needed for this sample project, better to keep it minimal.

diptadas replied:
Removed

Joerger commented on Design.md:
```diff
@@ -0,0 +1,128 @@
+
+- Simple RBAC style authorization
+- Permissions:
+    - Read and write: Required for creating, stopping, and purging a job
```
Are permissions tied to specific jobs? Or can a Read+Write user stop/query/purge any job?

diptadas replied:
No. They are tied with API usage.

Joerger commented on Design.md:
```diff
@@ -0,0 +1,128 @@
+- The go routine should exit when job finishes or timeout reached or force stop requested
+
+## Library
+
```
Can you add return types to each of these functions?

awly replied:
Let's keep the design doc high-level.
We can sort out function signatures during code review.
The important things here are: how are jobs tracked, where does the output go, what failures are possible (fail to start, preemptive stop, etc)

Joerger commented on Design.md:
```diff
@@ -0,0 +1,128 @@
+    - Initialize a job object using the ID
+    - Save the reference to the job into memory for future access
+    - Start the job in a separate go routine
+    - Return the job object
```
Can CreateJob return an error?

diptadas replied:
No. In the current design, it will return an aknowldegmet immediately with the Job ID and `CREATED` status. Any startup error or execution error can be checked by querying the `status` API.

Joerger replied:
Similarly to awly's comment below on `StopJob`, startup errors should be sent in the `StartJob` response. Otherwise users may need to check the status after each Start request to make sure it is running. This change may make the `Created` status unnecessary.

Joerger commented on Design.md:
```diff
@@ -0,0 +1,128 @@
+
+- [spf13/cobra](https://github.com/spf13/cobra) for CLI
+- [gorilla/mux](https://github.com/gorilla/mux) for routing
+- [mikespook/gorbac](github.com/mikespook/gorbac) for simple RBAC implementation
```
This RBAC should be simple enough to use plain go rather than adding a dependency.

diptadas replied:
Thank you for the suggestion. Updated the design doc.

awly commented on Design.md:
```diff
@@ -0,0 +1,128 @@
+## Library
+
+- `CreateJob(command, timeout)`
+    - Generate a unique ID
```
how will the ID be generated?

diptadas replied:
Using the [rs/xid](github.com/rs/xid) library we can generate a globally unique ID.

awly commented on Design.md:
```diff
@@ -0,0 +1,128 @@
+## Job object
+
+- `ID` unique identifier of a job
+- `Command` linux command to be executed
```
are command arguments passed separately?
for example, when a user runs `echo "foo bar"`, how are the arguments passed?

diptadas replied:
Users can specify the arguments as a list of strings.

diptadas committed:
Update design doc

diptadas commented:
Changelog for [ca1b72a](https://github.com/diptadas/job-worker/pull/1/commits/ca1b72aeb3be71adb0e120b8607f2962347dc753):

- Removed `purge` API
- Removed `timeout`
- Don't wrap the command within `bash`, instead use the original command
- Added arguments along with the command. Since we are not wrapping the command within `bash`, we can't use a single string to represent both command and arguments. Instead, users should provide a list of strings as arguments.
- Specified types for each field of the job object
- Added sample request and response for API endpoints
- Simplified the RBAC implementation. Initially, I thought of two layers of mapping: username to roles and roles to permissions. That's why I used the library. We can use only one mapping: username to permissions and get rid of the library.

awly commented on Design.md:
```diff
@@ -0,0 +1,161 @@
+## Job status
+
+- `Created` Received a job creation request (short-lived since we immediately start the job instead of queueing)
+- `Running` Job process started
+- `Running_Force_Stop` Failed to kill the process after force stop requested
```
this is effectively the same status - the job is still running

awly commented on Design.md:
```diff
@@ -0,0 +1,161 @@
+
+- Start the process using `exec.Cmd` in a separate go-routine
+- Utilize go-channels to communicate with the go-routine
+- The go-routine should exit when the job finishes or timeout reached or force stop requested
```
```suggestion
- The go-routine should exit when the job finishes or force stop requested
```

awly commented on Design.md:
```diff
@@ -0,0 +1,161 @@
+- `StopJob(jobID)`
+    - Fetch the job object from memory by job ID
+    - Get the current status of the job and check if it is running or already stopped
+    - Stop the job process if it is running using `Cmd.Process.Kill()`. However, there can be error while terminating the process and it can be still running. Use `Running_Force_Stop` status to indicate this scenario.
```
If `Kill()` fails, return an error in the `StopJob` response but keep the status as `Running`

awly commented on Design.md:
````diff
@@ -0,0 +1,161 @@
+
+Response:
+```json
+{
+  "message": "Termination requested for job c2l1qbucie6hpdufnung, please check job status"
+}
````
nit: this response doesn't seem useful
if `StopJob` succeeds, respond with `200 OK` and no body

awly commented on Design.md:
```diff
@@ -0,0 +1,161 @@
+
+- Simplified RBAC style authorization
+- Permissions:
+    - Read and write: Required for creating, stopping, and purging a job
```
```suggestion
    - Read and write: Required for creating and stopping a job
```

awly commented on Design.md:
```diff
@@ -0,0 +1,161 @@
+- `Exited_Error` Job finished with error
+- `Exited_Force_Stop` Job exited after force stop requested
+
+## Job handling
```
where is the job output stored?

awly commented on Design.md:
```diff
@@ -0,0 +1,161 @@
+
+- `job-worker [server/client]` combined CLI for running server and client
+- `job-worker server [--port] [--cacert] [--cert] [--key]` run the API server
+- `job-worker client [create/stop/purge/status] [--address] [--cacert] [--cert] [--key]` run client to communicate with the API server
+- `job-worker client create [--command] [--timeout]` request a new job
```
```suggestion
- `job-worker client [create/stop/status] [--address] [--cacert] [--cert] [--key]` run client to communicate with the API server
- `job-worker client create [--command]` request a new job
```

nklaassen reviewed (approved):
LGTM once @awly's current comments are addressed.

Joerger commented on Design.md:
```diff
@@ -0,0 +1,161 @@
+- `Arguments []string` list of arguments for the command
+- `Status string` current status of the job
+- `Output string` output of the job process (combined stderr and stdout)
+- `Error string` startup error, termination error, or exit error
```
```suggestion
- `Error error` startup error, termination error, or exit error
```

Joerger commented on Design.md:
```diff
@@ -0,0 +1,161 @@
+- `Running` Job process started
+- `Running_Force_Stop` Failed to kill the process after force stop requested
+- `Exited` Job finished successfully
+- `Exited_Error` Job finished with error
```
nit: This is the same as `j.status=Exited && j.error != nil`

diptadas committed:
Simplified job life cycle

diptadas commented:
Changes:
- `CreateJob` and `StopJob` will return error on failure to start/stop
- Removed `Created`, `Running_Force_Stop`, `Exited_Error` status
- Clarified how the output will be collected and processed

@awly @Joerger @nklaassen

awly commented on Design.md:
```diff
@@ -0,0 +1,160 @@
+- `Command string` Linux command to be executed
+- `Arguments []string` list of arguments for the command
+- `Status string` current status of the job
+- `OutputBuffer bytes.Buffer` links stderr and stdout of the job process, it will be hidden from API response
```
nit: you will need some synchronization around this, if you want to forward both stdout and stderr to the same buffer
